Rework link functions (#259)
Genentech/jmpost: vignettes/extending-jmpost.Rmd

Please note that this document is currently a work-in-progress and does not
contain complete information for this package yet. 

## Custom Link Functions

> Users can defined custom link functions in several ways based upon the level of customisation

Reviewer: ```suggestion
Users can define custom link functions in several ways based upon the level of customisation
```

$$

Where:
- $X$ is the design matrix of covariates
- $\beta$ is the vector of coefficients for the covariates
- $f(t)$ and $g(t)$ are the link functions
- $\alpha_1$ and $\alpha_2$ are the coefficients for the link functions
- $h_0(t)$ is the baseline hazard function
- $\phi_i$ is a vector of parameters from the longitudinal model

Each longitudinal model is responsible for defining their own implementations of the $\phi$ vector.
The interface for doing this is by providing an `enableLink` method that updates the `StanModule`
object of the model to define a Stan matrix with the name `link_function_inputs`
that contains 1 row per
patient and 1 column per $\phi$ parameter.

For reference the following is roughly the implementation for the `LongitudinalGSF` model:

```R
enableLink.LongitudinalGSF <- function(object, ...) {
  stan <- StanModule("
transformed parameters {
    matrix[Nind, 4] link_function_inputs;
    link_function_inputs[,1] = lm_gsf_psi_bsld;
    link_function_inputs[,2] = lm_gsf_psi_ks;
    link_function_inputs[,3] = lm_gsf_psi_kg;
    link_function_inputs[,4] = lm_gsf_psi_phi;
}")
    object@stan <- merge(
        object@stan,
        stan
    )

>     object

Reviewer: side comment (not for this PR): it might be nice to have an `update(object, stan)` method that does this, in a bit more readable and concise way

  key = "link_dsld",
  stan = StanModule("
functions {
    matrix link_dsld_contrib(
        matrix time,
        matrix link_function_inputs
    ) {
        int nrows = rows(time);
        int ncols = cols(time);
        vector[nrows] lm_rs_ind_rnd_slope = link_function_inputs[,2];
        matrix[nrows, ncols] rnd_slope_mat = rep_matrix(lm_rs_ind_rnd_slope, ncols);
        return rnd_slope_mat;
    }
}"),
  parameters = ParameterList(Parameter(name = "link_dsld", prior = prior, size = 1))
)
```

You can then pass these `LinkComponent` objects to the `link` argument of the `JointModel`
constructor to add them to the model. If you wish to add multiple link functions then you must
wrap them in a `Link()` object e.g.

```R
Link(
  LinkComponent(...),
  LinkComponent(...)
)
```

Note that there are a few families of link functions that are common across all models for example
identity, dSLD, TTG. Users can access these by simply using the inbuilt `link_identity()`,
`link_dsld()` and `link_ttg()` functions respectively.
These functions are responsible for loading the correct `LinkComponent` to implement that link for a
particular model.

If you wish to provide implementations of these

> for new models then you need to provide method implementations for your model to `linkDSLD()`,

Reviewer: maybe worth commenting on the difference of the snake case functions above and these camel case versions here

Reviewer: or is that something to fix here?

Note that there are a few families of link functions that are common across all models for example
identity, dSLD, TTG. Users can access these by simply using the inbuilt `link_identity()`,
`link_dsld()` and `link_ttg()` functions respectively.
These functions are responsible for loading the correct `LinkComponent` to
implement that link for a particular model.
That is if the user wants to specify both the
TTG link and the dSLD for a model then they can do so via:

```R
JointModel(
  ...,
  link = Link(
    link_dsld(),
    link_ttg()
  )
) 
```

Under the hood these link-family functions work by defining a `LinkComponent` object
that specifies a method that will return a `StanModule` object instead of specifying
the `StanModule` object directly. This method will then be called by the `JointModel` constructor
with the `LongitudinalModel` object as its first argument. This way each longitudinal model is 

> able to specify their own implementation for the given link family. The following is rough outline

Reviewer: ```suggestion
able to specify their own implementation for the given link family. The following is a rough outline
```

for how this works for the `link_dsld()` function and the `LongitudinalGSF` model:

```R
link_dsld <- function() {
  LinkComponent(
    key = "link_dsld",
    stan = linkDSLD,
    parameters = ParameterList(Parameter(name = "link_dsld", prior = prior, size = 1))
  )
}

linkDSLD.LongitudinalGSF <- function(object, ...) {
  StanModule("<stan code for GSF dSLD link>")
}
```

> This is to say that if you wish to overwrite a models implementation for existing link family or

Reviewer: ```suggestion
This is to say that if you wish to overwrite a model implementation for an existing link family or
```

> provide an implementaiton for a new model then you can do so by defining/overwriting the relevant

Reviewer: ```suggestion
provide an implementation for a new model then you can do so by defining/overwriting the relevant
```
